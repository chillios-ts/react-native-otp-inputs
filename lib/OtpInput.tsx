import React, { ReactElement, Component } from 'react'
import { TextInput, View, StyleSheet } from 'react-native'

interface Props {
  containerStyles?: any
  error?: boolean
  focusedBorderColor?: string
  handleBackspace: (event) => void
  inputStyles?: any
  textErrorColor?: string
  unFocusedBorderColor: string
  updateText: (text) => void
  value?: string
}

interface State {
  isFocused: boolean
}

export default class OtpInput extends Component<Props, State> {
  state = {
    isFocused: false,
  }

  private input: ReactElement<TextInput>

  public render() {
    const {
      containerStyles,
      error,
      focusedBorderColor,
      handleBackspace,
      inputStyles,
      textErrorColor,
      unFocusedBorderColor,
      updateText,
      value,
    } = this.props

    return (
      <View
        style={[
          defaultStyles.container,
          containerStyles,
          { borderColor: this.state.isFocused ? focusedBorderColor : unFocusedBorderColor },
        ]}
      >
        <TextInput
          clearTextOnFocus={true}
          keyboardType="phone-pad"
          maxLength={1}
          onBlur={this._onBlur}
          onChangeText={updateText}
          onFocus={this._onFocus}
          onKeyPress={handleBackspace}
          ref={input => (this.input = input as any)}
          selectTextOnFocus={true}
          style={[defaultStyles.otpInput, inputStyles, error && { color: textErrorColor }]}
          underlineColorAndroid="transparent"
          value={value}
        />
      </View>
    )
  }

  private _onFocus = () => this.setState({ isFocused: true })
  private _onBlur = () => this.setState({ isFocused: false })
}

const defaultStyles = StyleSheet.create({
  container: {
    backgroundColor: '#a1a1a1',
    borderRadius: 6,
    borderWidth: 1,
    height: 53,
    margin: 10,
  },
  otpInput: {
    color: '#ffffff',
    fontSize: 24,
    paddingTop: 10,
    textAlign: 'center',
    width: 40,
  },
})
