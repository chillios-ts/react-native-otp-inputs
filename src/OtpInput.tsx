import React, { useState, useEffect, forwardRef } from 'react';
import {
  Platform,
  StyleProp,
  TextInput,
  TextStyle,
  View,
  ViewStyle,
  TextInputProps,
} from 'react-native';

type Props = TextInputProps & {
  inputContainerStyles?: StyleProp<ViewStyle>;
  firstInput: boolean;
  focusStyles?: StyleProp<ViewStyle>;
  inputStyles?: StyleProp<TextStyle>;
  numberOfInputs: number;
  handleTextChange: (text: string) => void;
};

const majorVersionIOS: number = parseInt(`${Platform.Version}`, 10);
const isOTPSupported: boolean = Platform.OS === 'ios' && majorVersionIOS >= 12;

const OtpInput = forwardRef<TextInput, Props>(
  (
    {
      firstInput,
      focusStyles,
      handleTextChange,
      inputContainerStyles,
      inputStyles,
      numberOfInputs,
      placeholder,
      selectTextOnFocus,
      value,
      ...rest
    },
    ref,
  ) => {
    const [focused, setFocused] = useState(false);

    useEffect(() => {
      // @ts-ignore
      ref.current.setNativeProps({ value, text: value });
    }, [ref, value]);

    return (
      <View style={[inputContainerStyles, focused && focusStyles]}>
        <TextInput
          maxLength={firstInput ? numberOfInputs : 1}
          onBlur={() => setFocused(false)}
          onChangeText={handleTextChange}
          onFocus={() => setFocused(true)}
          ref={ref}
          placeholder={placeholder}
          // https://github.com/facebook/react-native/issues/18339
          selectTextOnFocus={Platform.select({ ios: selectTextOnFocus, android: true })}
          style={inputStyles}
          textContentType={isOTPSupported ? 'oneTimeCode' : 'none'}
          underlineColorAndroid="transparent"
          {...rest}
        />
      </View>
    );
  },
);

export default OtpInput;
