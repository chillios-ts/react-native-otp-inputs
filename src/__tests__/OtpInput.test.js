import { TextInput, Platform } from 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import OtpInput from '../OtpInput'

jest.mock('Platform', () => ({
  OS: 'ios',
  Version: 12,
}))

describe('<OtpInput />', () => {
  defaultProps = {
    handleBackspace: jest.fn(),
    updateText: jest.fn(),
  }

  describe('render', () => {
    test('with default props', () => {
      const wrapper = renderer.create(<OtpInput {...defaultProps} />)

      expect(wrapper).toMatchSnapshot()
    })

    test('with error message', () => {
      const wrapper = renderer.create(
        <OtpInput {...defaultProps} error={true} errorMessage="Error" textErrorColor="#00ff00" />,
      )

      expect(wrapper).toMatchSnapshot()
    })

    test('with iOS version 12 or higher', () => {
      const wrapper = renderer.create(<OtpInput {...defaultProps} />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('_onFocus', () => {
    test('should change state of `isFocused` to true', () => {
      const wrapperInstance = renderer.create(<OtpInput {...defaultProps} />).getInstance()
      wrapperInstance.setState({ isFocused: false })
      wrapperInstance._onFocus()

      expect(wrapperInstance.state.isFocused).toEqual(true)
    })
  })

  describe('_onBlur', () => {
    test('should change state of `isFocused` to false', () => {
      const wrapperInstance = renderer.create(<OtpInput {...defaultProps} />).getInstance()
      wrapperInstance.setState({ isFocused: true })
      wrapperInstance._onBlur()

      expect(wrapperInstance.state.isFocused).toEqual(false)
    })
  })
})
