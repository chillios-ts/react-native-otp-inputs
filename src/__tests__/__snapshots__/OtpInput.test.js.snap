// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OtpInput /> render with default props 1`] = `
<View
  style={
    Array [
      Object {
        "borderBottomWidth": 1,
        "height": 53,
        "margin": 10,
      },
      undefined,
      false,
      Object {
        "borderColor": undefined,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    maxLength={1}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "fontSize": 24,
          "paddingTop": 10,
          "textAlign": "center",
          "width": 40,
        },
        undefined,
        undefined,
      ]
    }
    textContentType="none"
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`<OtpInput /> render with error message 1`] = `
<View
  style={
    Array [
      Object {
        "borderBottomWidth": 1,
        "height": 53,
        "margin": 10,
      },
      undefined,
      false,
      Object {
        "borderColor": undefined,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    maxLength={1}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "fontSize": 24,
          "paddingTop": 10,
          "textAlign": "center",
          "width": 40,
        },
        undefined,
        Object {
          "color": "#00ff00",
        },
      ]
    }
    textContentType="none"
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`<OtpInput /> render with iOS version 12 or higher 1`] = `
<View
  style={
    Array [
      Object {
        "borderBottomWidth": 1,
        "height": 53,
        "margin": 10,
      },
      undefined,
      false,
      Object {
        "borderColor": undefined,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    maxLength={1}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "fontSize": 24,
          "paddingTop": 10,
          "textAlign": "center",
          "width": 40,
        },
        undefined,
        undefined,
      ]
    }
    textContentType="none"
    underlineColorAndroid="transparent"
  />
</View>
`;
